src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,196 @@
+        newNode.next = nodeToInsertAfter;
+        nodeToInsertAfter.prev.next = newNode;
+        nodeToInsertAfter.prev = newNode;
+    }
+
+    public void printLinkedList() {
+        Node currentNode = head;
+
+        while (currentNode != null) {
+            System.out.println(currentNode.data);
+            currentNode = currentNode.next;
+        }
+    }
+
+    public void printReversedLinkedList() {
+        Node currentNode = tail;
+
+        while (currentNode != null) {
+            System.out.println(currentNode.data);
+            currentNode = currentNode.prev;
+        }
+    }
+
+    private Node unlink(Node node) {
+        size--;
+        if (node == tail) {
+            tail = node.prev;
+            return node;
+        }
+        if (node == head) {
+            head = node.next;
+            return node;
+        }
+        node.prev.next = node.next;
+        node.next.prev = node.prev;
+        return node;
+    }
+
     @Override
     public void add(T value) {
+        add(value, size);
     }
 
     @Override
     public void add(T value, int index) {
+        validateIndexExclusive(index);
+        Node newNode = new Node(value);
+        if (isEmpty()) {
+            head = newNode;
+            tail = newNode;
+        } else if (index == 0) {
+            newNode.next = head;
+            head.prev = newNode;
+            head = newNode;
+        } else if (index == size) {
+            newNode.prev = tail;
+            tail.next = newNode;
+            tail = newNode;
+        } else {
+            Node currentNode = getNode(index);
+            insertBefore(currentNode, newNode);
+        }
+        size++;
+        System.out.println("------------");
+        printLinkedList();
+        System.out.println("------------");
+        printReversedLinkedList();
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNode(index).data;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node nodeToUpdate = getNode(index);
+        T oldValue = nodeToUpdate.data;
+        nodeToUpdate.data = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node nodeToRemove = getNode(index);
+        return unlink(nodeToRemove).data;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        int index = getIndexOfNode(object);
+        if (index == -1) {
+            return false;
+        }
+        remove(index);
+        return true;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    class Node {
```

Comment:
implementation of our list should be hidden from outwhere
```suggestion
   private class Node {
```